### DISCUSSION TITLE
単体テストの考え方/使い方 6章

### DISCUSSION THREAD

**Discussion by nmurakami0** [atman-inc/training#101] [General]
輪読会のサマリ用ディスカッション。
6章分の各章単位程度でコメントを記載するので、不明点、相談点、確認点、勉強になったポイントなどは、各自ディスカッションのスレッドへコメントをお願いします。

**Comment by nmurakami0** [comment]
## 5.4 振り返り: 単体テストの古典学派とロンドン学派の違い

### 5.4.1 モックに置き換える必要のないプロセス外依存

### 5.4.2 モックを使った振る舞いの検証

**Reply by nmurakami0** [reply]
- 古典学派では共有依存を単体テストで使わないことを推奨
- 共有依存がプロセス外依存なら、通常はテストダブルに置き換えてテストする
    - ただ、プロセス外依存なら何でもモックに置き換えないといけないわけではない
        - もしプロセス外依存にアクセスするのにテスト対象のアプリケーションを必ず経由させなくてはならないようになっているならば、テスト対象のアプリケーションとそのプロセス外依存とのコミュニケーションはシステムの観察可能な振る舞いの一部にはならない
            - -> アプリケーションを経由するということは、アプリケーションとそのプロセス外依存とのコミュニケーションが実装の詳細になっているということだから
                - 実装の詳細なので、その場合は後方互換を気にする必要はない
                - -> e.g., ORMを通してしかDBにアクセスしないなら、たとえばDBではなく別のデータストレージに置き換えても問題ない
                - プロセス外依存をモックに置き換えてしまうとリファクタリングへの耐性がなくなってしまう
- モックは振る舞いを検証するものとして見られることがあるが、モックするのは実装の詳細であり観察可能な振る舞いではないため、ここで見る振る舞いは粒度が細かすぎるものとなっている
    - -> 例えばAPIのロジックの一部としてWeb APIを叩く場合に、その部分のモックは何をリクエストにしてどのようなレスポンスを得るかという振る舞いを検証する
        - これは観察可能な振る舞いではない

**Reply by koushisa** [reply]
- 観察可能な振る舞いと依存の種類
  - 引用
    - システム内コミュニケーション」と「システム間コミュニケーション」は区別すべき
    - 外部から観測できないプロセス外依存とのコミュニケーションは実装の詳細になる
    - ヘキサゴナルアーキテクチャでドメイン層とアプリケーションサービス層の2層に分離できる
  - 意訳
    - 観測可能な振る舞いと実装詳細は区別すべし
    - データと振る舞いはなるべく近い位置に配置して凝集度を高めてスコープは小さく、カプセルすべし
    - 凝集度高めるのも大事だけどデータ構造と振る舞いを切り離すと後方互換性も維持しやすい
  - 感じたこと
    - 共有依存を扱う際のテクニックとして知られるDIやlocal reasoningなども、依存の主体定義を親ではなく末端に閉じさせることで疎結合にする & 後方互換性を維持するというのが本質っぽい
    - 依存制御、関心の分離、カプセル化の技法は言語やフレームワークごとに異なる(むしろほとんどがフレームワークに依存する)ので、思想に精通してないと難しいところはある

**Reply by nmurakami0** [reply]
memo

- rspec-parallelみたいなのを使うときは、並列数分のデータベースを先に作っておく
  - 共有依存が起きないように

**Comment by nmurakami0** [comment]
# 第6章 単体テストの3つの手法

## 6.1 単体テストの3つの手法

### 6.1.1 出力値ベース・テストとは？

**Reply by nmurakami0** [reply]
- 出力値ベース・テスト
    - 副作用がない場合のみ可能

**Reply by xxManatoxx** [reply]
- 単体テストの3つの手法
  - 出力値ベース・テスト
    - 戻り値を確認するテスト
    - 確認対象のメソッドに引数などを渡してResultを確認するテストをイメージ
  - 状態ベース・テスト
    - 状態を確認するテスト
    - DBに登録・更新したデータが正しい件数で正しい内容で登録・更新されているかのテストをイメージ
  - コミュニケーション・ベース・テスト 
    - オブジェクト間のやり取りを確認するテスト
    - 読んでいる時点でイメージできず
      - オブジェクト間だと結合テストが思い浮かんでしまった
        - 先にこの部分の説明を読む
          - 協力者オブジェクトのモックを作り正しく呼び出せているかを確認している
- 感じたこと
  - 今までこの様に明確に状態なのか、戻り値なのか意識できていなかったですが、今後意識することでどのようなテストを行いたいかの整理が自分の中でしやすくなりそうだなと感じました。

**Comment by nmurakami0** [comment]
### 6.1.2 状態ベース・テストとは？

**Comment by nmurakami0** [comment]
### 6.1.3 コミュニケーションベース・テストとは？

**Reply by nmurakami0** [reply]
ロンドン学派はテストダブルを多様するのでコミュニケーションベーステストを好む

**Comment by nmurakami0** [comment]
## 6.2 単体テストの3つの手法の比較

**Reply by nmurakami0** [reply]
"regressionに対する保護"と"迅速なフィードバック"は、どのテスト手法を選択するかの影響を受けない

**Comment by y-kaneda** [comment]
### 6.2.1 退行に対する保護の観点での比較と迅速なフィードバックの観点での比較

**Reply by y-kaneda** [reply]
- 退行に対する保護は単体テストの手法によっては変わらない
- コミュニケーション・ベース・テスト(モックを用いる)だけは使い過ぎると検証されるプロダクション・コードの量が減ってしまう

**Reply by nmurakami0** [reply]
単体テストの手法はデグレの出やすさとは関係ないということ

**Comment by y-kaneda** [comment]
### 6.2.2 リファクタリングへの耐性の観点での比較

**Reply by y-kaneda** [reply]
- 偽陽性(false positive)がリファクタリングの際にどのくらい発生するか
  - ケースが観察可能な振る舞いではなく実装の詳細と結びつくことで発生する

- 出力値ベース・テストが最も偽陽性の発生を抑えられる
  - テスト対象メソッド単体だけで検証できる

- 状態ベース・テストはオブジェクトへの副作用も見ることになるのでテスト・ケースが実装の詳細と結びつきやすくなってしまう

- コミュニケーション・テストはリファクタリングへの耐性が低い。テスト対象の本筋と関係ないところを検証しているとリファクタリング時に壊れてしまう

**Reply by xxManatoxx** [reply]
- 出力値ベース・テスト
  - 偽陽性の発生を最も抑えられる
  - 実装の詳細と結びつく唯一のケース
    - テスト対象メソッド自体が実装の詳細の時
- 状態ベース・テスト
  - 出力ベース・テストと比べると偽陽性が発生しやすくなる
  - テストケースとプロダクションコードの結びつきが増えるほど偽陽性の機会が増える
- コミュニケーション・ベース・テスト
  - 偽陽性に対して最も脆弱

単体テストの手法によって注意する観点は変わってくる

**Reply by nmurakami0** [reply]
- リファクタリングへの耐性
  - false positive: "実装を変更した際にテストが間違ってfailしてしまう"ことが多いかなので、耐性の強さは以下の順になる
    - 出力値ベース > 状態ベース > コミュニケーションベース
      - 直感的には実装に結びつきやすいテストほど耐性がない

実装をうまくすれば、極力関数型っぽくすることは出来るので、そうすると出力値ベーステストにもなって、テスト的にもメリットが出るということ

**Comment by y-kaneda** [comment]
### 6.2.3 保守のしやすさの観点での比較

**Reply by y-kaneda** [reply]
- 保守のしやすさの目安
  - テスト・ケースの理解の難易度
    - テスト・ケースのコード量に相関
  - テスト・ケースの実施の難易度
    - テスト・ケースで扱うプロセス外依存の量に相関
 
- 出力値ベース・テストは入力値・出力値を気にするだけなので最も保守をしやすい
- 状態ベース・テストは状態を検証するため、多くの場合、テスト・ケースのコードが多くなる
  - 共通のヘルパー・メソッドを使って検証のコード量を減らすことはできる
    - ただし、テストのためだけにプロダクション・コードを値オブジェクトにする等は**プロダクション・コードの汚染**になる
- コミュニケーションベース・テストはテスト対象システムとテスト・ダブルのコミュニケーションを確認するためのコード量が増える。また**モックの連鎖**が起こるとさらにテスト・コードの量が増えてしまう

**Reply by xxManatoxx** [reply]
- テストケースを理解することがどのくらい難しいか
  - コード量が多いとテスト内容を理解したり、変更するのが難しくなる
    - 参画している案件で大きめのテストコードを見ることが多く簡単なテスト内容でも複雑に見えていた
    - 自分が実装するときにも気をつけて実装したい
- テストケースを実施するのがどのくらい難しいか
  - プロセス外依存の数が増えるとその依存の稼働が大変になる

- 出力値ベース・テスト
  - 通常はコード量が少なく、簡潔になるため保守しやすいテストになる
- 状態ベース・テスト
  - 確認フェーズが増える傾向にある
  - ヘルパーメソッドを作成してコード量を減らすこともできるがヘルパーメソッドの保守も考慮しなくてはいけなくなる
    - ここで作成したヘルパーメソッドも他のテストケースでも呼ばれない限り労力に対する価値として戻ってくることは無い
  - 値オブジェクトを利用して確認対象の比較を簡単に行えるようにする方法もある
    - 効果的ではあるが、確認対象が本質的に値である必要がある、かつ値オブジェクトに変換可能な場合のみしか使えない
    - 過度にテストを単純化する目的のためにプロダクションコードを変更するのはプロダクションコードの汚染することになる
- コミュニケーション・ベース・テスト
  - テストダブルの用意、テストダブルとテスト対象のシステムのコミュニケーションの確認の必要が出てくる
  - 上記から多くのテストコード実装が必要になってしまう為、保守がしにくくなる

**Reply by nmurakami0** [reply]
確かにテストは概して理解する、実施するのが難しいことが多い。

出力値ベース > 状態ベース > コミュニケーションベースの順で保守しやすい

- 状態ベースは値オブジェクトを使うとオブジェクトの各属性をテスト時に比較しないで単に値として比較出来るので、いくらか見やすくなる
- コミュニケーションベースは、テストダブルを用意しそのコミュニケーションをテストで確認しないといけない
  - しかもモックの連鎖になりがち

**Comment by y-kaneda** [comment]
### 6.2.4 単体テストの3つの手法の比較に関するまとめ

**Reply by y-kaneda** [reply]
- 出力値ベース・テストがもっともコスパが良い
  - 実装の詳細と結びつくことがほとんどない
  - プロセス外依存も扱わない
- 状態ベース・テストとコミュニケーション・ベース・テストは出力値ベーステストよりも劣る
  - 実装の詳細と結びつきやすい
  - 記述するコード量が増える

**Comment by y-kaneda** [comment]
## 6.3 関数型アーキテクチャについて

**Comment by y-kaneda** [comment]
### 6.3.1 関数型プログラミングとは?

**Reply by nmurakami0** [reply]
- 副作用
- 例外: 隠れた出力
- 内部もしくは外部の状態の参照: 隠れた入力

副作用はよく上げられるが、残り2つが意外に見逃しがちと思った。
DBからのデータ取得はわかるが、DateTime.Nowは言われてみると確かにという感じ。

**Reply by y-kaneda** [reply]
- 数学的関数は純粋関数とも呼ばれる
  - 隠れた入出力がない関数
    - 全てメソッドの引数・戻り値だけで明示される
    - 他の状態の影響を受けないため、冪等になる

- 隠れた入出力がない
  - 副作用
    - メソッド・シグネチャに表現されていない、隠れた出力
  - 例外
    - 入出力とは関係のない処理の流れであり、呼び出し・返却の流れを飛び越えてどこかでキャッチされることになる、隠れた出力
  - 内部もしくは外部の状態への参照
    - 直接パラメータで指定された以外の隠れた入力

**Reply by koushisa** [reply]
一休のCTO(伊藤直也さん)が解説していたGraphQLバックエンドは実践的な関数型になっていてわかりやすい
- [TypeScript による GraphQL バックエンド開発](https://speakerdeck.com/naoya/typescript-niyoru-graphql-batukuendokai-fa-75b3dab7-90a8-4169-a4dc-d1e7410b9dbd?slide=41)

**Comment by y-kaneda** [comment]
### 6.3.2 関数型アーキテクチャとは何か?

**Reply by nmurakami0** [reply]
関数型プログラミングが目標としていることは、副作用を完全に取り除くことではなく、ビジネスロジックを扱うコードと副作用を起こすコードを分離することにある

- 副作用をビジネス・オペレーションの最初や最後に持ってくることで、分離しやすくしている
  - この最初と最後という言葉は、図6.9を見るとわかりやすい
- functional core・immutable core: 数学的関数を使って書くことが出来る部分
- mutable shell: 数学的関数によって下された決定を観測可能な振る舞いの一部に変換する
  - "観測可能な振る舞いの一部に変換する"部分が副作用
- 単体テストではfunctional coreを出力値ベーステストで検証し、mutable shellの検証は、統合テストに任せる
- 関数型アーキテクチャは"アーキテクチャ"と名がつくように全体的な視点で、immutabilityは部分的視点
  - immutabilityとカプセル化は、目指すところは近い
    - immutableにしてしまえば、初期化後に変更出来ないのでカプセル化漏れで生じる問題が起こり得ない

感想

- 大昔にSparkを触っていたが、Scalaで実装していたので言語としては勿論関数型だが、意図せず関数型アーキテクチャにもなっていた様子
  - 分散ノードでデータの一部が処理されるので、データが集約されてmainノードに来るまで副作用を発生させることが物理的に実用的でないため
  - とはいえ分散データベースにアクセスするジョブもあったので、その場合は関数型アーキテクチャが崩れているかも

**Reply by y-kaneda** [reply]
目標：ビジネスロジックを扱うコードと副作用を起こすコードを分離する

- 副作用をビジネス・オペレーションの最初や最後に持っていく
  - 入力や関数の結果の使用や保存等
- 決定を下すコード
  - 副作用を起こさない
  - 数学的関数を用いる
  - 関数的核(functional core)、不変核(immutable core)
- 決定に基づくアクションを実行するコード
  - 数学的関数によって下された決定を観察可能な振る舞いの一部に変換する
    - DB操作、メッセージ送信等
  - 可変殻(mutable shell)
  - 入力もこれに当たり、関数的核へ渡す

- 関数的核の部分は出力値ベース・テストで行い、可変殻の検証は統合テストに任せる

- 関数型における不変性とカプセル化は目的は同じ
  - カプセル化は対象オブジェクトを不変オブジェクトにして正しい状態が崩れないようにし、関数型は最初だけオブジェクトの状態を確認すればあとは冪等になるので正しい状態に維持される

**Comment by y-kaneda** [comment]
### 6.3.3 関数型アーキテクチャとヘキサゴナル・アーキテクチャの比較

**Reply by nmurakami0** [reply]
- 関数型アーキテクチャはヘキサゴナルアーキテクチャの一種
  - 副作用の扱いが異なる
    - ヘキサゴナルアーキテクチャでは、ドメイン層内なら副作用を起こすことが許されている
      - e.g., インスタンスの状態を変えることが許されている

**Reply by y-kaneda** [reply]
- ともに関心の分離(separation of concerns)が基盤
  - 責務の分離
  - 依存の流れが一方向
- 副作用の扱いで差がある
  - 関数型アーキテクチャは全ての副作用を関数的核の外に出す
  - ヘキサゴナル・アーキテクチャではドメイン層内に収まっていれば副作用が許される
    - インスタンス自体の状態を変えることは許されている

**Comment by y-kaneda** [comment]
## 6.4 関数型アーキテクチャおよび出力値ベース・テストへの移行

**Comment by y-kaneda** [comment]
### 6.4.1 サンプル・プロジェクト(訪問者記録システム)について

**Reply by nmurakami0** [reply]
- 問題点: ドメイン層とアプリケーションサービス層が一つになっている
  - 具体的には、AuditManagerは、ファイルへの書き込みの仕方のロジックと、実際に書き込むところを同じAddRecordで行っている

**Reply by y-kaneda** [reply]
- ファイルに書き込みたい内容、書き込み方に関わる処理と、実際のファイル・ディレクトリ関連の確認・書き込み処理が密接になってしまっている
- テストをしようとすると
  - 実ファイルが共有依存になってしまい複数のテスト・ケースを同時に実行できない
  - 実ファイルが扱える状態であることを担保しておかなければならない

**Comment by y-kaneda** [comment]
### 6.4.2 モックを用いることによる単体テストとファイル・システムの分離

**Reply by nmurakami0** [reply]
- おさらい: 入力を置き換えるのがスタブ、出力を置き換えるのがモック

**Reply by y-kaneda** [reply]
- 実際のファイル・システム操作を分離し、依存性注入(DI)するようにする
- ファイル・システム操作の依存をモックに置き換える
- 想定されたモックのメソッドが正しく実行されていることで振る舞いの検証とする
- 実ファイル関連の準備・確認や複数テスト・ケースからの取り合い等も無くなる

**Reply by xxManatoxx** [reply]
モックを利用した単体テストに改修する

- プロセス外依存をモックに置き換える
- モックに置き換えられるように、FileSystemの箇所をInterfaceとして抽出
- Interfaceを注入されるように訪問者記録システムを改修する
- 共有依存となっていたファイルシステムがテスト間で共有されなくなり以下が改善される
  - 迅速なフィードバック
  - 保守のしやすさ

**Comment by y-kaneda** [comment]
### 6.4.3 関数型アーキテクチャへのリファクタリング

**Reply by y-kaneda** [reply]
- ファイルの情報を受け取り、書き込む内容と書き込み方を返す(決定だけをする)関数にする
  - 決定をするだけの関数になると、副作用を起こさない
- 決定された内容を受け取り副作用を起こす関数を独立する
- 入力・決定・出力を指揮するアプリケーション・サービス

**Comment by y-kaneda** [comment]
### 6.4.4 開発がさらに進んだ時のシナリオ

**Comment by y-kaneda** [comment]
## 6.5 関数型アーキテクチャの欠点

**Reply by y-kaneda** [reply]
- メリット
  - テストのしやすさ
  - 保守のしやすさ
- デメリット
  - 導入の難易度
  - パフォーマンス劣化
  - コードベースの肥大化

**Comment by y-kaneda** [comment]
### 6.5.1 関数型アーキテクチャの導入が難しい場合

**Reply by y-kaneda** [reply]
- 最初に全部入力を集約できる場合はうまいこと行く
- 決定の途中結果から改めてデータを取得する必要がある場合等は難しくなる
  - 使うか分からないデータを先に取得しておく
    - パフォーマンスが低下
  - データが必要になるかどうかをあらかじめ判定し、必要であれば取得しておく
    - 責務の分離が損なわれる

**Comment by y-kaneda** [comment]
### 6.5.2 パフォーマンスに関する欠点

**Reply by y-kaneda** [reply]
- 全体的にプロセス外依存への呼び出しが増える
  - 入力データの収集 -> 決定 -> 実行 の流れになるため

**Reply by xxManatoxx** [reply]
- 関数型アーキテクチャを選ぶのか、伝統的なアーキテクチャを選ぶのか
  - →パフォーマンスを選ぶのか、実装コードの保守コストを選ぶのかの選択になる
- アーキテクチャはシステムが叶えたい要望に合わせて考慮するべきで必ずしも万能なアーキテクチャは存在しない


- 考えたこと
  - アーキテクチャの最適解の選択は難しいと思うので、以下行っていきたい
    - アーキテクチャの知識をつける
    - 自分が携わるシステムの要望をよく理解して考えられるように経験を積んでいく

**Comment by y-kaneda** [comment]
### 6.5.3 コードベースが大きくなってしまう欠点

**Reply by y-kaneda** [reply]
- functional core と mutable shell を明確に分離するためにコード量が増える
- 複雑、重要なことをしないシステムの場合、そのデメリットがメリットを上回る
- 原理主義的に採用するのではなく、できるだけ多くの出力値ベース・テストができるようなイメージ

**Comment by y-kaneda** [comment]
### まとめ

**Reply by y-kaneda** [reply]
- 副作用を減らして出力値ベース・テストの比率を上げていくのが良い
